pkg/common/scheduler/state/state.go

```diff
@@ -0,0 +1,399 @@
+		return zoneName, nodeName, err
+	}
+
+	nodeName = pod.Spec.NodeName
+	zoneName, ok := s.NodeToZoneMap[nodeName]
+	if !ok {
+		return zoneName, nodeName, errors.New("could not find zone")
+	}
+	return zoneName, nodeName, nil
+}
+
+// stateBuilder reconstruct the state from scratch, by listing vpods
+type stateBuilder struct {
+	ctx               context.Context
+	logger            *zap.SugaredLogger
+	vpodLister        scheduler.VPodLister
+	capacity          int32
+	schedulerPolicy   scheduler.SchedulerPolicyType
+	nodeLister        corev1.NodeLister
+	statefulSetClient clientappsv1.StatefulSetInterface
+	statefulSetName   string
+	podLister         corev1.PodNamespaceLister
+	schedPolicy       *scheduler.SchedulerPolicy
+	deschedPolicy     *scheduler.SchedulerPolicy
+}
+
+// NewStateBuilder returns a StateAccessor recreating the state from scratch each time it is requested
+func NewStateBuilder(ctx context.Context, namespace, sfsname string, lister scheduler.VPodLister, podCapacity int32, schedulerPolicy scheduler.SchedulerPolicyType, schedPolicy *scheduler.SchedulerPolicy, deschedPolicy *scheduler.SchedulerPolicy, podlister corev1.PodNamespaceLister, nodeLister corev1.NodeLister) StateAccessor {
+
+	return &stateBuilder{
+		ctx:               ctx,
+		logger:            logging.FromContext(ctx),
+		vpodLister:        lister,
+		capacity:          podCapacity,
+		schedulerPolicy:   schedulerPolicy,
+		nodeLister:        nodeLister,
+		statefulSetClient: kubeclient.Get(ctx).AppsV1().StatefulSets(namespace),
+		statefulSetName:   sfsname,
+		podLister:         podlister,
+		schedPolicy:       schedPolicy,
+		deschedPolicy:     deschedPolicy,
+	}
+}
+
+func (s *stateBuilder) State(reserved map[types.NamespacedName]map[string]int32) (*State, error) {
+	vpods, err := s.vpodLister()
+	if err != nil {
+		return nil, err
+	}
+
+	scale, err := s.statefulSetClient.GetScale(s.ctx, s.statefulSetName, metav1.GetOptions{})
+	if err != nil {
+		s.logger.Infow("failed to get statefulset", zap.Error(err))
+		return nil, err
+	}
+
+	free := make([]int32, 0)
+	schedulablePods := make([]int32, 0)
+	last := int32(-1)
+
+	// keep track of (vpod key, podname) pairs with existing placements
+	withPlacement := make(map[types.NamespacedName]map[string]bool)
+
+	podSpread := make(map[types.NamespacedName]map[string]int32)
+	nodeSpread := make(map[types.NamespacedName]map[string]int32)
+	zoneSpread := make(map[types.NamespacedName]map[string]int32)
+
+	noexec := &v1.Taint{
+		Key:    "node.kubernetes.io/unreachable",
+		Effect: v1.TaintEffectNoExecute,
+	}
+
+	nosched := &v1.Taint{
+		Key:    "node.kubernetes.io/unreachable",
+		Effect: v1.TaintEffectNoSchedule,
+	}
+
+	//Build the node to zone map
+	nodes, err := s.nodeLister.List(labels.Everything())
+	if err != nil {
+		return nil, err
+	}
+
+	nodeToZoneMap := make(map[string]string)
+	zoneMap := make(map[string]struct{})
+	for i := 0; i < len(nodes); i++ {
+		node := nodes[i]
+		if node.Spec.Unschedulable {
```

Comment:
I think this line requires some explanation. In theory the vpod scheduler shouldn't have to deal with nodes directly. A pod on an unschedulable node might still be viable, right?

Reply:
Thin line I feel. A node is marked as Unschedulable in its Spec after it has been optionally drained and cordoned. K8 scheduler won't place new pods onto that Node, so idea was that we _should not_ place "new vreplicas" onto this pod either. Because the next step for this Node might be rebooting/shutting down/upgrading etc. So this is a preventative step to add more rebalancing work. That was my thinking..